### PR TITLE
Links 

### PR DESCRIPTION
**Pull request by rkoster**
This PR addresses: https://github.com/cloudfoundry-incubator/haproxy-boshrelease/issues/55.
I would have liked this PR to be more generic, however I was unable to find a way to consume links with conflicting link types.

### CODE REVIEW

File: jobs/haproxy/spec [hunk 1]
```diff
@@ -165,6 +174,8 @@ properties:
                                  # Setting `verify` requires `ha_proxy.backend_ca_file` key to be set.
           backend_verifyhost: example.com # optional - hostname to verify in the x509 certificate subject for SSL-enabled backend servers
                                           # only used if backend_ssl: `verify` is set
+  ha_proxy.tcp_link_port:
+    description: "Port haproxy should listen on when using the tcp_backend link"
```
Comment (geofffranks) [inline jobs/haproxy/spec]: Shouldn't this be something we pull from the link itself?

**Comment by rkoster** [comment]
I would love to, however at this point the diego release is not exposing this property:
https://github.com/cloudfoundry/diego-release/blob/develop/jobs/ssh_proxy/spec#L13-L15 <https://github.com/cloudfoundry/diego-release/blob/develop/jobs/ssh_proxy/spec#L13-L15>
I have tried specifying link properties (http://bosh.io/docs/links-manual.html <http://bosh.io/docs/links-manual.html>) but this is not allowed when also doing normal linking..

**Comment by geofffranks** [comment]
Yeah, I hadn't yet gotten to the part where you try to use the link then fail-over. It's all good.

Though looking at the bosh.io <http://bosh.io/> docs, the way you're calling p() isn't mentioned (to provide a default if not present). I assume it works, since you were testing with diego not providing the property, right?

**Comment by rkoster** [comment]
@geofffranks [here](https://github.com/cloudfoundry/bosh/blob/master/src/bosh-template/lib/bosh/template/evaluation_context.rb#L65-L70) you can find defaults for `p()`.

**Comment by geofffranks** [comment]
Thanks!

**Comment by geofffranks** [comment]
Soon as release notes get through CI

**Comment by rkoster** [comment]
nice!

**Comment by ematpl** [comment]
Hey, @geofffranks, I'm a little surprised that you accepted this PR with the `tcp_backend` link as type `ssh_proxy` instead of as a more generic identifier, given that it's still essentially specific to the [`ssh_proxy` job in diego-release](https://github.com/cloudfoundry/diego-release/blob/v1.22.0/jobs/ssh_proxy/spec#L13-L15). I'd be open to having that job in diego-release provide a separate link type (say, `tcp-backend`) if it allows the HAProxy release to be more generic. Along those lines, it also seems like it would be beneficial to supply the backend port in the link: the template is looking for it at the reasonable location of `port`, but unfortunately the `ssh_proxy` job embeds that port in its [`diego.ssh_proxy.listen_addr` property](https://github.com/cloudfoundry/diego-release/blob/v1.22.0/jobs/ssh_proxy/spec#L25-L27), and BOSH links don't yet offer a mechanism to rename or otherwise manipulate properties in the providing job to different properties in the link. That said, I'm starting to consider how to de-namespace and simplify the diego-release job properties now that BOSH recommends scoping the properties in a manifest to each job in an instance group, and it would be reasonable there to extract that port into a separate property (say, `port` or `server_port`) that would be more natural to include in the links that job provides.  

Also pinging @cppforlife and @Amit-PivotalLabs for any suggestions they may have on link design and/or potential forthcoming link capabilities.

Thanks again,
Eric

**Pull request merged**